account-info/src/lib.rs

```diff
@@ -130,23 +130,18 @@ impl<'a> AccountInfo<'a> {
             .map_err(|_| ProgramError::AccountBorrowFailed)
     }
 
-    /// Realloc the account's data and optionally zero-initialize the new
-    /// memory.
+    /// Resize the account's data: Either truncating or zero extending.
     ///
     /// Note:  Account data can be increased within a single call by up to
     /// `solana_program::entrypoint::MAX_PERMITTED_DATA_INCREASE` bytes.
     ///
-    /// Note: `zero_init` being `false` will no longer be supported by the
-    /// program runtime.
-    ///
     /// # Safety
     ///
     /// This method makes assumptions about the layout and location of memory
     /// referenced by `AccountInfo` fields. It should only be called for
     /// instances of `AccountInfo` that were created by the runtime and received
     /// in the `process_instruction` entrypoint of a program.
-    #[deprecated(since = "2.3.0", note = "Use AccountInfo::resize() instead")]
-    pub fn realloc(&self, new_len: usize, zero_init: bool) -> Result<(), ProgramError> {
+    pub fn resize(&self, new_len: usize) -> Result<(), ProgramError> {
```

Comment:
I am in two minds on this one. My understanding is that this behaviour will be introduced with ABIv2, where to resize will be done via a syscall. It will most likely require a different `AccountInfo`-like implementation.

If we mimic the same behaviour now, it will cost more to resize an account since we are always zeroing the data. At the same time, the current behaviour will continue to be supported so in a way it is unnecessary to zero the data.

What do you think if we keep the deprecated so people are aware that changes are coming but still allow the current `realloc` to be used?

Reply:
Ah ok, I thought this was going to be broken much sooner, ie before ABIv2. @Lichtso what do you think? If it won't be broken soon, then we can definitely keep it around for longer.

Reply:
@febo This will come in the global direct mapping feature, before the opt-in ABIV2.

Reply:
Oh, how is it going to work for existing programs that expect the area to be zeroed? E.g., will loader v3 stop zeroing the account data "padding"?

Reply:
All loaders will stop zeroing the padding. Programs have to write to it first to use it (which can be done using the memset syscall or normal store instructions).

Reply:
Isn't there a risk that it will break existing programs? For example, a program that resizes an account and then loads a type using a zero-copy library (e.g., bytemuck) assuming that the memory is zeroed.

Reply:
Given that this behaviour is going to change sooner than ABIv2 and the fact that `realloc` should really not be used anymore, I agree that removing it is the best way forward.

Reply:
In that case, can I get a :heavy_check_mark: ?